core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/EvaluateComparison.java

```diff
@@ -0,0 +1,87 @@
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.IsNull;
+import io.trino.sql.ir.optimizer.IrOptimizerRule;
+import io.trino.sql.planner.Symbol;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.sql.ir.Booleans.NULL_BOOLEAN;
+import static io.trino.sql.ir.Comparison.Operator.EQUAL;
+import static io.trino.sql.ir.Comparison.Operator.GREATER_THAN;
+import static io.trino.sql.ir.Comparison.Operator.GREATER_THAN_OR_EQUAL;
+import static io.trino.sql.ir.Comparison.Operator.IDENTICAL;
+import static io.trino.sql.ir.Comparison.Operator.LESS_THAN;
+import static io.trino.sql.ir.Comparison.Operator.LESS_THAN_OR_EQUAL;
+import static io.trino.sql.ir.Comparison.Operator.NOT_EQUAL;
+import static io.trino.sql.ir.IrExpressions.isConstantNull;
+
+public class EvaluateComparison
+        implements IrOptimizerRule
+{
+    private final Metadata metadata;
+    private final InterpretedFunctionInvoker functionInvoker;
+
+    public EvaluateComparison(PlannerContext context)
+    {
+        metadata = context.getMetadata();
+        functionInvoker = new InterpretedFunctionInvoker(context.getFunctionManager());
+    }
+
+    @Override
+    public Optional<Expression> apply(Expression expression, Session session, Map<Symbol, Expression> bindings)
+    {
+        return switch (expression) {
+            case Comparison(Operator operator, Constant left, Constant right) when operator == IDENTICAL -> Optional.of(evaluate(OperatorType.IDENTICAL, left, right, session));
+            case Comparison(Operator operator, Constant left, Constant right) when operator == EQUAL -> Optional.of(evaluate(OperatorType.EQUAL, left, right, session));
+            case Comparison(Operator operator, Constant left, Constant right) when operator == NOT_EQUAL -> Optional.of(switch (evaluate(OperatorType.EQUAL, left, right, session)) {
+                case Constant constant when isConstantNull(constant) -> constant;
+                case Constant(Type type, Boolean value) -> new Constant(type, !value);
+                default -> throw new IllegalStateException("Unexpected value for boolean expression");
+            });
+            case Comparison(Operator operator, Constant left, Constant right) when operator == LESS_THAN -> Optional.of(evaluate(OperatorType.LESS_THAN, left, right, session));
+            case Comparison(Operator operator, Constant left, Constant right) when operator == LESS_THAN_OR_EQUAL -> Optional.of(evaluate(OperatorType.LESS_THAN_OR_EQUAL, left, right, session));
+            case Comparison(Operator operator, Constant left, Constant right) when operator == GREATER_THAN -> Optional.of(evaluate(OperatorType.LESS_THAN, right, left, session));
+            case Comparison(Operator operator, Constant left, Constant right) when operator == GREATER_THAN_OR_EQUAL -> Optional.of(evaluate(OperatorType.LESS_THAN_OR_EQUAL, right, left, session));
```

Comment:
Tried that, but it muddles things a bit. The nested switch makes the code harder to read. It also requires adding a synthetic catch-all case in the nested switch since it doesn't cover all possible operators.